module/zfs/zfs_ioctl.c

```diff
@@ -3258,22 +3311,30 @@ zfs_ioc_clone(const char *fsname, nvlist_t *innvl, nvlist_t *outnvl)
 {
 	int error = 0;
 	nvlist_t *nvprops = NULL;
+	nvlist_t *hidden_args = NULL;
+	dsl_crypto_params_t *dcp = NULL;
 	char *origin_name;
 
 	if (nvlist_lookup_string(innvl, "origin", &origin_name) != 0)
 		return (SET_ERROR(EINVAL));
 	(void) nvlist_lookup_nvlist(innvl, "props", &nvprops);
+	(void) nvlist_lookup_nvlist(innvl, ZPOOL_HIDDEN_ARGS, &hidden_args);
```

Comment:
The atomicity is nice.  The same could be said of all the properties.  Do you think the atomicity of the crypto properties is more important enough to justify the special case code in dmu_objset_clone() and callees? (Personally I'm undecided - just something to think about.)

Hopefully some day we can clean this up with channel programs, combining the clone, change-key, and prop setting into one atomic channel program.